File: js/algorithms/binary-search.js

function search(target, stack, callback) {
    var min     = stack[0];
    var max     = stack[stack.length - 1];
    var guess   = 0;
    var guesses = 0;

    while (true) {
        guess = Math.floor((min + max) / 2);

        if (stack[guess] === target) {
            return callback(null, {target: target, value: stack[guess], key: guess});
        }

        if (target > stack[guess]) {
            min = guess + 1;
        }
        else {
            max = guess - 1;
        }
    }
}

function range(to) {
    var range = [];

    for (var i = 0; i <= to; i++) {
        range.push(i);
    }

    return range;
}


search(
    20,
    range(50000000),
    function(err, result) {
        console.log(result);
    }
);

File: c/algorithms/binary-search.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int search(int target, int stack[]);
int * range(int to);

int main(void) {
    int *r = range(200);

    printf("We found key %d for value %d searched.\n", search(200, r), 200);

    free(r);

    exit(0);
}

int * range(int to) {
    static int *range;

    range = malloc(to * sizeof(int));

    for (int i = 0; i <= to; i++) {
        range[i] = i + 1;
    }

    return range;
}

int search(int target, int *stack) {
    int min = stack[0];
    int max = stack[19999999];

    int guess;

    while (1) {
        guess = floor((min + max) / 2);

        if (stack[guess] == target) {
            return guess;
        }

        if (target > stack[guess]) {
            min = guess + 1;
        }
        else {
            max = guess - 1;
        }
    }
}
